client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java

```diff
@@ -150,6 +150,8 @@ private PositionInternal getPosition() {
     @GuardedBy("readers")
     private String updateGroupStateIfNeeded() throws ReinitializationRequiredException {
         try {
+            // join StreamCut Barrier and update state with latest positions.
+            groupState.checkAndJoinStreamCutBarrier(getPosition().getOwnedSegmentsWithOffsets());
```

Comment:
Joining the stream cut barrier is being done automatically upon calling `readNextEvent` and implicitly this method. We might want to expose a method for the application to join a specific reader. Let's use the quarter example to illustrate. We can't determine in Pravega that a given segment has received the first event for a given quarter, that requires application logic.